From df253a64aba96d79e86b750de9304ac61a728389 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 10 Aug 2026 10:24:55 +0700
Subject: [PATCH 242/243] Revert "usbcore: Add quirk for 255-bytes initial
 config read"

This reverts commit bdc6a40fbde61c90dfcfc2af4317451afb32922a.

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 .../admin-guide/kernel-parameters.txt         | 10 -----
 drivers/usb/core/config.c                     | 39 ++++---------------
 drivers/usb/core/quirks.c                     |  4 --
 include/linux/usb/quirks.h                    |  3 --
 4 files changed, 7 insertions(+), 49 deletions(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 14121458a0c0..b5493a7f8f22 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -8169,16 +8169,6 @@ Kernel parameters
 				q = USB_QUIRK_FORCE_ONE_CONFIG (Device
 					claims zero configurations,
 					forcing to 1);
-				r = USB_QUIRK_WINDOWS_CONFIG_REQ_SIZE (Device
-					fails during initialization when asked for
-					9-bytes configuration descriptor request.
-					Ask for 255-bytes request instead to mirror
-					Windows' behavior. This quirk is originally
-					meant to fix some quirky gamepads that refuse
-					to connect in their XInput mode. But it can
-					also potentially fix issues with other USB
-					devices that work on Windows but not on
-					Linux);
 			Example: quirks=0781:5580:bk,0a5c:5834:gij
 
 	usbhid.mousepoll=
diff --git a/drivers/usb/core/config.c b/drivers/usb/core/config.c
index 442c15f92ccd..45e20c6d76c0 100644
--- a/drivers/usb/core/config.c
+++ b/drivers/usb/core/config.c
@@ -912,17 +912,6 @@ int usb_get_configuration(struct usb_device *dev)
 	unsigned char *bigbuffer;
 	struct usb_config_descriptor *desc;
 	int result;
-	size_t usb_config_req_size;
-
-	/*
-	 * Devices with quirky firmware will stall or reset when the initial
-	 * config descriptor request uses wLength=9. If the quirk is set, use
-	 * 255 instead, mirroring the behavior of Windows.
-	 */
-	if (dev->quirks & USB_QUIRK_WINDOWS_CONFIG_REQ_SIZE)
-		usb_config_req_size = 255;
-	else
-		usb_config_req_size = USB_DT_CONFIG_SIZE;
 
 	if (ncfg > USB_MAXCONFIG) {
 		dev_notice(ddev, "too many configurations: %d, "
@@ -949,19 +938,15 @@ int usb_get_configuration(struct usb_device *dev)
 	if (!dev->rawdescriptors)
 		return -ENOMEM;
 
-	desc = kmalloc(usb_config_req_size, GFP_KERNEL);
+	desc = kmalloc(USB_DT_CONFIG_SIZE, GFP_KERNEL);
 	if (!desc)
 		return -ENOMEM;
 
 	for (cfgno = 0; cfgno < ncfg; cfgno++) {
-		/*
-		 * Normally we request only the configuration descriptor header
-		 * so we can determine the total configuration length. For
-		 * devices with USB_QUIRK_WINDOWS_CONFIG_REQ_SIZE set, try to
-		 * grab the full descriptor set instead.
-		 */
+		/* We grab just the first descriptor so we know how long
+		 * the whole configuration is */
 		result = usb_get_descriptor(dev, USB_DT_CONFIG, cfgno,
-		    desc, usb_config_req_size);
+		    desc, USB_DT_CONFIG_SIZE);
 		if (result < 0) {
 			dev_err(ddev, "unable to read config index %d "
 			    "descriptor/%s: %d\n", cfgno, "start", result);
@@ -971,8 +956,9 @@ int usb_get_configuration(struct usb_device *dev)
 			dev->descriptor.bNumConfigurations = cfgno;
 			break;
 		} else if (result < 4) {
-			dev_err(ddev, "config index %d descriptor too short (asked for %zu, got %i, need at least %i)\n",
-			    cfgno, usb_config_req_size, result, 4);
+			dev_err(ddev, "config index %d descriptor too short "
+			    "(expected %i, got %i)\n", cfgno,
+			    USB_DT_CONFIG_SIZE, result);
 			result = -EINVAL;
 			goto err;
 		}
@@ -986,16 +972,6 @@ int usb_get_configuration(struct usb_device *dev)
 			goto err;
 		}
 
-		/*
-		 * If the device returns the full configuration descriptor set,
-		 * skip the second read. Otherwise, send a second request
-		 * asking for the full set.
-		 */
-		if (result >= length) {
-			memcpy(bigbuffer, desc, length);
-			goto store_and_parse;
-		}
-
 		if (dev->quirks & USB_QUIRK_DELAY_INIT)
 			msleep(200);
 
@@ -1013,7 +989,6 @@ int usb_get_configuration(struct usb_device *dev)
 			length = result;
 		}
 
-store_and_parse:
 		dev->rawdescriptors[cfgno] = bigbuffer;
 
 		result = usb_parse_configuration(dev, cfgno,
diff --git a/drivers/usb/core/quirks.c b/drivers/usb/core/quirks.c
index f5a60ccf21d3..87ee2d938bc0 100644
--- a/drivers/usb/core/quirks.c
+++ b/drivers/usb/core/quirks.c
@@ -142,10 +142,6 @@ static int quirks_param_set(const char *value, const struct kernel_param *kp)
 				break;
 			case 'q':
 				flags |= USB_QUIRK_FORCE_ONE_CONFIG;
-				break;
-			case 'r':
-				flags |= USB_QUIRK_WINDOWS_CONFIG_REQ_SIZE;
-				break;
 			/* Ignore unrecognized flag characters */
 			}
 		}
diff --git a/include/linux/usb/quirks.h b/include/linux/usb/quirks.h
index a4043b33c2c2..b3cc7beab4a3 100644
--- a/include/linux/usb/quirks.h
+++ b/include/linux/usb/quirks.h
@@ -81,7 +81,4 @@
 /* Device claims zero configurations, forcing to 1 */
 #define USB_QUIRK_FORCE_ONE_CONFIG		BIT(18)
 
-/* Use a 255 bytes config descriptor request mirroring windows behavior */
-#define USB_QUIRK_WINDOWS_CONFIG_REQ_SIZE	BIT(19)
-
 #endif /* __LINUX_USB_QUIRKS_H */
-- 
2.55.0

